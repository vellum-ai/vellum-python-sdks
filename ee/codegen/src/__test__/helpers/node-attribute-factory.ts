import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
