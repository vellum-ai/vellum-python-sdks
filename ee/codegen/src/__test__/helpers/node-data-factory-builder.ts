import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
