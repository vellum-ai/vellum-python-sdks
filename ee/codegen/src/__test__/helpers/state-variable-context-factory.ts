import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
