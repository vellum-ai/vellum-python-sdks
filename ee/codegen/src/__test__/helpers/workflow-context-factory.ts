import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
