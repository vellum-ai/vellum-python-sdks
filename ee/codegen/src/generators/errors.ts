export type CodegenErrorCode =
  | "PROJECT_SERIALIZATION_ERROR"
  | "WORKFLOW_GENERATION_ERROR"
  | "WORKFLOW_INPUT_GENERATION_ERROR"
  | "NODE_DEFINITION_GENERATION_ERROR"
  | "NODE_ATTRIBUTE_GENERATION_ERROR"
  | "NODE_PORT_GENERATION_ERROR"
  | "NODE_NOT_FOUND_ERROR"
  | "NODE_OUTPUT_NOT_FOUND_ERROR"
  | "NODE_INPUT_NOT_FOUND_ERROR"
  | "UNSUPPORTED_SANDBOX_INPUT_ERROR"
  | "ENTITY_NOT_FOUND_ERROR"
  | "POST_PROCESSING_ERROR"
  | "VALUE_GENERATION_ERROR";

export abstract class BaseCodegenError extends Error {
  abstract code: CodegenErrorCode;
}

/**
 * An error that raises when deserializing the request to codegen
 * into a valid Workflow project.
 */
export class ProjectSerializationError extends BaseCodegenError {
  code = "PROJECT_SERIALIZATION_ERROR" as const;
}

/**
 * An error that raises when the Workflow definition fails to
 * generate.
 */
export class WorkflowGenerationError extends BaseCodegenError {
  code = "WORKFLOW_GENERATION_ERROR" as const;
}

/**
 * An error that raises when the Workflow Inpus fail to
 * generate.
 */
export class WorkflowInputGenerationError extends BaseCodegenError {
  code = "WORKFLOW_INPUT_GENERATION_ERROR" as const;
}

/**
 * An error that raises when the Node definition fails to generate
 */
export class NodeDefinitionGenerationError extends BaseCodegenError {
  code = "NODE_DEFINITION_GENERATION_ERROR" as const;
}

/**
 * An error that raises when generating a node attribute fails.
 */
export class NodeAttributeGenerationError extends BaseCodegenError {
  code = "NODE_ATTRIBUTE_GENERATION_ERROR" as const;
}

/**
 * An error that raises when generating a node port fails.
 */
export class NodePortGenerationError extends BaseCodegenError {
  code = "NODE_PORT_GENERATION_ERROR" as const;
}

/**
 * An error that raises when a node is not found.
 */
export class NodeNotFoundError extends BaseCodegenError {
  code = "NODE_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a vellum entity is not found.
 */
export class EntityNotFoundError extends BaseCodegenError {
  code = "ENTITY_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a sandbox input is not supported.
 */
export class UnsupportedSandboxInputError extends BaseCodegenError {
  code = "UNSUPPORTED_SANDBOX_INPUT_ERROR" as const;
}

/**
 * An error that raises when a node output is not found.
 */
export class NodeOutputNotFoundError extends BaseCodegenError {
  code = "NODE_OUTPUT_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when a node input is not found.
 */
export class NodeInputNotFoundError extends BaseCodegenError {
  code = "NODE_INPUT_NOT_FOUND_ERROR" as const;
}

/**
 * An error that raises when the code fails to post process
 */
export class PostProcessingError extends BaseCodegenError {
  code = "POST_PROCESSING_ERROR" as const;
}

/**
 * An error that raises when a value fails to generate
 */
export class ValueGenerationError extends BaseCodegenError {
  code = "VALUE_GENERATION_ERROR" as const;
}
