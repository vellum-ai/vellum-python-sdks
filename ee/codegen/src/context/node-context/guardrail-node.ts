import { MetricDefinitionHistoryItem, VellumVariableType } from "vellum-ai/api";
import { MetricDefinitions as MetricDefinitionsClient } from "vellum-ai/api/resources/metricDefinitions/client/Client";
import { VellumError } from "vellum-ai/errors";

import { BaseNodeContext } from "./base";

import { PortContext } from "src/context/port-context";
import { EntityNotFoundError } from "src/generators/errors";
import { GuardrailNode as GuardrailNodeType } from "src/types/vellum";

export class GuardrailNodeContext extends BaseNodeContext<GuardrailNodeType> {
  baseNodeClassName = "GuardrailNode";
  baseNodeDisplayClassName = "BaseGuardrailNodeDisplay";

  public metricDefinitionsHistoryItem: MetricDefinitionHistoryItem | undefined =
    undefined;

  getNodeOutputNamesById(): Record<string, string> {
    if (!this.metricDefinitionsHistoryItem) {
      return {};
    }

    return this.metricDefinitionsHistoryItem.outputVariables.reduce(
      (acc, variable) => {
        acc[variable.id] = variable.key;
        return acc;
      },
      {} as Record<string, string>
    );
  }

  getNodeOutputTypesById(): Record<string, VellumVariableType> {
    if (!this.metricDefinitionsHistoryItem) {
      return {};
    }

    return Object.fromEntries(
      this.metricDefinitionsHistoryItem.outputVariables.map((variable) => [
        variable.id,
        variable.type,
      ])
    );
  }

  createPortContexts(): PortContext[] {
    return [
      new PortContext({
        workflowContext: this.workflowContext,
        nodeContext: this,
        portId: this.nodeData.data.sourceHandleId,
      }),
    ];
  }

  async buildProperties(): Promise<void> {
    let metricDefinitionsHistoryItem: MetricDefinitionHistoryItem | undefined =
      undefined;

    try {
      metricDefinitionsHistoryItem = await new MetricDefinitionsClient({
        apiKey: this.workflowContext.vellumApiKey,
        environment: this.workflowContext.vellumApiEnvironment,
      }).metricDefinitionHistoryItemRetrieve(
        this.nodeData.data.releaseTag,
        this.nodeData.data.metricDefinitionId
      );
    } catch (e) {
      if (e instanceof VellumError && e.statusCode === 404) {
        this.workflowContext.addError(
          new EntityNotFoundError(
            `Metric Definition "${this.nodeData.data.metricDefinitionId} ${this.nodeData.data.releaseTag}" not found.`,
            "WARNING"
          )
        );
      } else {
        throw e;
      }
    }

    this.metricDefinitionsHistoryItem = metricDefinitionsHistoryItem;
  }
}
