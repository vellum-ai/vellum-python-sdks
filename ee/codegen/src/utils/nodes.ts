import { VellumError } from "vellum-ai/errors";

import { WorkflowNode } from "src/types/vellum";

export function getNodeLabel(nodeData: WorkflowNode): string {
  switch (nodeData.type) {
    case "GENERIC":
      return nodeData.definition?.name ?? nodeData.label ?? "Generic Node";
    default:
      return nodeData.data.label;
  }
}

export function isUnaryOperator(operator: string): boolean {
  return operator === "null" || operator === "notNull";
}

export function isVellumErrorWithDetail(
  error: unknown
): error is VellumError & { body: { detail: string } } {
  return (
    error instanceof VellumError &&
    typeof error.body === "object" &&
    error.body !== null &&
    "detail" in error.body &&
    typeof error.body.detail === "string"
  );
}
