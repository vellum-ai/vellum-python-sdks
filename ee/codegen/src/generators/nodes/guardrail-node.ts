import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import { BaseNode } from "./bases/base";

import { OUTPUTS_CLASS_NAME } from "src/constants";
import { GuardrailNodeContext } from "src/context/node-context/guardrail-node";
import { NodeAttributeGenerationError } from "src/generators/errors";
import { GuardrailNode as GuardrailNodeType } from "src/types/vellum";

const INPUTS_PREFIX = "metric_inputs";
const STANDARD_METRIC_OUTPUT_KEYS = ["score", "normalized_score", "log"];

export class GuardrailNode extends BaseNode<
  GuardrailNodeType,
  GuardrailNodeContext
> {
  protected DEFAULT_TRIGGER = "AWAIT_ANY";
  protected getNodeAttributeNameByNodeInputKey(nodeInputKey: string): string {
    return `${INPUTS_PREFIX}.${nodeInputKey}`;
  }

  getNodeClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    if (!this.nodeData.data.metricDefinitionId) {
      throw new NodeAttributeGenerationError(
        "metric_definition_id is required"
      );
    }

    statements.push(
      python.field({
        name: "metric_definition",
        initializer: python.TypeInstantiation.str(
          this.nodeData.data.metricDefinitionId
        ),
      })
    );

    statements.push(
      python.field({
        name: INPUTS_PREFIX,
        initializer: python.TypeInstantiation.dict(
          Array.from(this.nodeInputsByKey.entries()).map(([key, value]) => ({
            key: python.TypeInstantiation.str(key),
            value: value,
          })),
          {
            endWithComma: true,
          }
        ),
      })
    );

    if (!this.nodeData.data.releaseTag) {
      throw new NodeAttributeGenerationError("release_tag is required");
    }

    statements.push(
      python.field({
        name: "release_tag",
        initializer: python.TypeInstantiation.str(
          this.nodeData.data.releaseTag
        ),
      })
    );

    return statements;
  }

  getNodeDisplayClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    statements.push(
      python.field({
        name: "target_handle_id",
        initializer: python.TypeInstantiation.uuid(
          this.nodeData.data.targetHandleId
        ),
      })
    );

    return statements;
  }

  protected getOutputDisplay(): python.Field {
    return python.field({
      name: "output_display",
      initializer: python.TypeInstantiation.dict(
        (
          this.nodeContext.metricDefinitionsHistoryItem?.outputVariables ?? []
        ).map((output) => {
          const name = this.nodeContext.getNodeOutputNameById(output.id);
          const isStandardOutput = STANDARD_METRIC_OUTPUT_KEYS.includes(
            output.key
          );

          if (!name) {
            throw new NodeAttributeGenerationError(
              `Could not find output name for ${this.nodeContext.nodeClassName}.Outputs.${output.key} given output id ${output.id}`
            );
          }

          if (isStandardOutput) {
            // For standard outputs, use class reference
            return {
              key: python.reference({
                name: this.nodeContext.nodeClassName,
                modulePath: this.nodeContext.nodeModulePath,
                attribute: [OUTPUTS_CLASS_NAME, name],
              }),
              value: python.instantiateClass({
                classReference: python.reference({
                  name: "NodeOutputDisplay",
                  modulePath:
                    this.workflowContext.sdkModulePathNames
                      .NODE_DISPLAY_TYPES_MODULE_PATH,
                }),
                arguments_: [
                  python.methodArgument({
                    name: "id",
                    value: python.TypeInstantiation.uuid(output.id),
                  }),
                  python.methodArgument({
                    name: "name",
                    value: python.TypeInstantiation.str(output.key),
                  }),
                ],
              }),
            };
          } else {
            // For non standard outputs, use a LazyReference
            return {
              key: python.instantiateClass({
                classReference: python.reference({
                  name: "LazyReference",
                  modulePath: [
                    ...this.workflowContext.sdkModulePathNames
                      .WORKFLOWS_MODULE_PATH,
                    "references",
                  ],
                }),
                arguments_: [
                  python.methodArgument({
                    value: python.TypeInstantiation.str(
                      `${this.nodeContext.nodeClassName}.${OUTPUTS_CLASS_NAME}.${output.key}`
                    ),
                  }),
                ],
              }),
              value: python.instantiateClass({
                classReference: python.reference({
                  name: "NodeOutputDisplay",
                  modulePath:
                    this.workflowContext.sdkModulePathNames
                      .NODE_DISPLAY_TYPES_MODULE_PATH,
                }),
                arguments_: [
                  python.methodArgument({
                    name: "id",
                    value: python.TypeInstantiation.uuid(output.id),
                  }),
                  python.methodArgument({
                    name: "name",
                    value: python.TypeInstantiation.str(output.key),
                  }),
                ],
              }),
            };
          }
        })
      ),
    });
  }

  protected getErrorOutputId(): string | undefined {
    return this.nodeData.data.errorOutputId;
  }
}
