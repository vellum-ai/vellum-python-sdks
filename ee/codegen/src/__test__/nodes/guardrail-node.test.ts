import { Writer } from "@fern-api/python-ast/core/Writer";
import { MetricDefinitionHistoryItem } from "vellum-ai/api";
import { MetricDefinitions as MetricDefinitionsClient } from "vellum-ai/api/resources/metricDefinitions/client/Client";
import { VellumError } from "vellum-ai/errors";
import { beforeEach, vi } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import { guardrailNodeDataFactory } from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { GuardrailNodeContext } from "src/context/node-context/guardrail-node";
import { GuardrailNode } from "src/generators/nodes/guardrail-node";

describe("GuardrailNode", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "a6ef8809-346e-469c-beed-2e5c4e9844c5",
          key: "expected",
          type: "STRING",
        },
        workflowContext,
      })
    );

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "abed55ada-923e-46ef-8340-1a5b0b563dc1",
          key: "actual",
          type: "STRING",
        },
        workflowContext,
      })
    );

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "1472503c-1662-4da9-beb9-73026be90c68",
          key: "output",
          type: "STRING",
        },
        workflowContext,
      })
    );
  });

  const mockMetricDefinition = (
    outputVariables: { id: string; key: string; type: string }[]
  ) => ({
    id: "mocked-metric-output-id",
    label: "mocked-metric-output-label",
    name: "mocked-metric-output-name",
    description: "mocked-metric-output-description",
    outputVariables,
  });

  const createNode = async ({
    errorOutputId,
    outputVariables,
  }: {
    errorOutputId?: string;
    outputVariables: { id: string; key: string; type: string }[];
  }) => {
    vi.spyOn(
      MetricDefinitionsClient.prototype,
      "metricDefinitionHistoryItemRetrieve"
    ).mockResolvedValue(
      mockMetricDefinition(
        outputVariables
      ) as unknown as MetricDefinitionHistoryItem
    );
    const nodeData = guardrailNodeDataFactory({
      errorOutputId,
      inputs: [
        {
          id: "3f917af8-03a4-4ca4-8d40-fa662417fe9c",
          key: "expected",
          value: {
            rules: [
              {
                type: "INPUT_VARIABLE",
                data: {
                  inputVariableId: "a6ef8809-346e-469c-beed-2e5c4e9844c5",
                },
              },
            ],
            combinator: "OR",
          },
        },
        {
          id: "bed55ada-923e-46ef-8340-1a5b0b563dc1",
          key: "actual",
          value: {
            rules: [
              {
                type: "INPUT_VARIABLE",
                data: {
                  inputVariableId: "1472503c-1662-4da9-beb9-73026be90c68",
                },
              },
            ],
            combinator: "OR",
          },
        },
      ],
    }).build();

    const nodeContext = (await createNodeContext({
      workflowContext,
      nodeData,
    })) as GuardrailNodeContext;

    return new GuardrailNode({
      workflowContext,
      nodeContext,
    });
  };

  describe("basic", () => {
    it.each([
      [
        "single output variable",
        [{ id: "mocked-input-id", key: "score", type: "NUMBER" }],
      ],
      [
        "multiple output variables",
        [
          { id: "mocked-input-id-1", key: "score1", type: "NUMBER" },
          { id: "mocked-input-id-2", key: "score2", type: "NUMBER" },
        ],
      ],
    ])("getNodeFile - %s", async (_, outputVariables) => {
      const node = await createNode({ outputVariables });

      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it.each([
      [
        "single output variable",
        [{ id: "mocked-input-id", key: "score", type: "NUMBER" }],
      ],
      [
        "multiple output variables",
        [
          { id: "mocked-input-id-1", key: "score1", type: "NUMBER" },
          { id: "mocked-input-id-2", key: "score2", type: "NUMBER" },
        ],
      ],
    ])("getNodeDisplayFile - %s", async (_, outputVariables) => {
      const node = await createNode({ outputVariables });

      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("reject on error enabled", () => {
    it("getNodeFile", async () => {
      const node = await createNode({
        errorOutputId: "38361ff1-c826-49b8-aa8d-28179c3684cc",
        outputVariables: [
          { id: "mocked-input-id-1", key: "score1", type: "NUMBER" },
          { id: "mocked-input-id-2", key: "score2", type: "NUMBER" },
        ],
      });

      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("getNodeDisplayFile", async () => {
      const node = await createNode({
        errorOutputId: "38361ff1-c826-49b8-aa8d-28179c3684cc",
        outputVariables: [
          { id: "mocked-input-id-1", key: "score1", type: "NUMBER" },
          { id: "mocked-input-id-2", key: "score2", type: "NUMBER" },
        ],
      });

      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("no metric definition found", () => {
    let node: GuardrailNode;
    beforeEach(async () => {
      vi.spyOn(
        MetricDefinitionsClient.prototype,
        "metricDefinitionHistoryItemRetrieve"
      ).mockRejectedValue(
        new VellumError({
          message: "Metric Definition not found",
          body: {
            detail: "Could not find Metric Definition",
          },
          statusCode: 404,
        })
      );
      writer = new Writer();
      workflowContext = workflowContextFactory({ strict: false });
      const nodeData = guardrailNodeDataFactory().build();

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as GuardrailNodeContext;

      node = new GuardrailNode({
        workflowContext,
        nodeContext,
      });
    });

    it(`getNodeFile`, async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();

      const errors = workflowContext.getErrors();
      // console.log(errors);
      expect(errors).toHaveLength(1);
      expect(errors[0]?.message).toContain(
        'Metric Definition "589df5bd-8c0d-4797-9a84-9598ecd043de LATEST" not found.'
      );
      expect(errors[0]?.severity).toBe("WARNING");
    });
  });
});
