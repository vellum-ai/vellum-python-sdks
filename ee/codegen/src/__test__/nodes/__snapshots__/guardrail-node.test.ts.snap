// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GuardrailNode > basic > getNodeDisplayFile - multiple output variables 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseGuardrailNodeDisplay
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.guardrail_node import GuardrailNode


class GuardrailNodeDisplay(BaseGuardrailNodeDisplay[GuardrailNode]):
    label = "Guardrail Node"
    node_id = UUID("metric")
    target_handle_id = UUID("1817fbab-db21-4219-8b34-0e150ce78887")
    metric_input_ids_by_name = {
        "expected": UUID("3f917af8-03a4-4ca4-8d40-fa662417fe9c"),
        "actual": UUID("bed55ada-923e-46ef-8340-1a5b0b563dc1"),
    }
    node_input_ids_by_name = {
        "expected": UUID("3f917af8-03a4-4ca4-8d40-fa662417fe9c"),
        "actual": UUID("bed55ada-923e-46ef-8340-1a5b0b563dc1"),
    }
    output_display = {
        GuardrailNode.Outputs.score1: NodeOutputDisplay(
            id=UUID("mocked-input-id-1"), name="score1"
        ),
        GuardrailNode.Outputs.score2: NodeOutputDisplay(
            id=UUID("mocked-input-id-2"), name="score2"
        ),
    }
    port_displays = {
        GuardrailNode.Ports.default: PortDisplayOverrides(
            id=UUID("92aafe31-101b-47d3-86f2-e261c7747c16")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=0, y=0), width=None, height=None
    )
"
`;

exports[`GuardrailNode > basic > getNodeDisplayFile - single output variable 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import BaseGuardrailNodeDisplay
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.guardrail_node import GuardrailNode


class GuardrailNodeDisplay(BaseGuardrailNodeDisplay[GuardrailNode]):
    label = "Guardrail Node"
    node_id = UUID("metric")
    target_handle_id = UUID("1817fbab-db21-4219-8b34-0e150ce78887")
    metric_input_ids_by_name = {
        "expected": UUID("3f917af8-03a4-4ca4-8d40-fa662417fe9c"),
        "actual": UUID("bed55ada-923e-46ef-8340-1a5b0b563dc1"),
    }
    node_input_ids_by_name = {
        "expected": UUID("3f917af8-03a4-4ca4-8d40-fa662417fe9c"),
        "actual": UUID("bed55ada-923e-46ef-8340-1a5b0b563dc1"),
    }
    output_display = {
        GuardrailNode.Outputs.score: NodeOutputDisplay(
            id=UUID("mocked-input-id"), name="score"
        )
    }
    port_displays = {
        GuardrailNode.Ports.default: PortDisplayOverrides(
            id=UUID("92aafe31-101b-47d3-86f2-e261c7747c16")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=0, y=0), width=None, height=None
    )
"
`;

exports[`GuardrailNode > basic > getNodeFile - multiple output variables 1`] = `
"from vellum.workflows.nodes.displayable import GuardrailNode as BaseGuardrailNode

from ..inputs import Inputs


class GuardrailNode(BaseGuardrailNode):
    metric_definition = "589df5bd-8c0d-4797-9a84-9598ecd043de"
    metric_inputs = {
        "expected": Inputs.expected,
        "actual": Inputs.output,
    }
    release_tag = "LATEST"
"
`;

exports[`GuardrailNode > basic > getNodeFile - single output variable 1`] = `
"from vellum.workflows.nodes.displayable import GuardrailNode as BaseGuardrailNode

from ..inputs import Inputs


class GuardrailNode(BaseGuardrailNode):
    metric_definition = "589df5bd-8c0d-4797-9a84-9598ecd043de"
    metric_inputs = {
        "expected": Inputs.expected,
        "actual": Inputs.output,
    }
    release_tag = "LATEST"
"
`;

exports[`GuardrailNode > reject on error enabled > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.nodes import (
    BaseGuardrailNodeDisplay,
    BaseTryNodeDisplay,
)
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)
from vellum_ee.workflows.display.vellum import NodeDisplayData, NodeDisplayPosition

from ...nodes.guardrail_node import GuardrailNode


@BaseTryNodeDisplay.wrap(error_output_id=UUID("38361ff1-c826-49b8-aa8d-28179c3684cc"))
class GuardrailNodeDisplay(BaseGuardrailNodeDisplay[GuardrailNode]):
    label = "Guardrail Node"
    node_id = UUID("metric")
    target_handle_id = UUID("1817fbab-db21-4219-8b34-0e150ce78887")
    metric_input_ids_by_name = {
        "expected": UUID("3f917af8-03a4-4ca4-8d40-fa662417fe9c"),
        "actual": UUID("bed55ada-923e-46ef-8340-1a5b0b563dc1"),
    }
    node_input_ids_by_name = {
        "expected": UUID("3f917af8-03a4-4ca4-8d40-fa662417fe9c"),
        "actual": UUID("bed55ada-923e-46ef-8340-1a5b0b563dc1"),
    }
    output_display = {
        GuardrailNode.Outputs.score1: NodeOutputDisplay(
            id=UUID("mocked-input-id-1"), name="score1"
        ),
        GuardrailNode.Outputs.score2: NodeOutputDisplay(
            id=UUID("mocked-input-id-2"), name="score2"
        ),
    }
    port_displays = {
        GuardrailNode.Ports.default: PortDisplayOverrides(
            id=UUID("92aafe31-101b-47d3-86f2-e261c7747c16")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=0, y=0), width=None, height=None
    )
"
`;

exports[`GuardrailNode > reject on error enabled > getNodeFile 1`] = `
"from vellum.workflows.nodes.core import TryNode
from vellum.workflows.nodes.displayable import GuardrailNode as BaseGuardrailNode

from ..inputs import Inputs


@TryNode.wrap()
class GuardrailNode(BaseGuardrailNode):
    metric_definition = "589df5bd-8c0d-4797-9a84-9598ecd043de"
    metric_inputs = {
        "expected": Inputs.expected,
        "actual": Inputs.output,
    }
    release_tag = "LATEST"
"
`;
